Brokerages/GDAX/GDAXBrokerage.Messaging.cs

```diff
@@ -185,6 +191,102 @@ private void OnMessageImpl(object sender, WebSocketMessage e)
             }
         }
 
+        private void OnSnapshot(string data)
+        {
+            try
+            {
+                var message = JsonConvert.DeserializeObject<Messages.Snapshot>(data);
+
+                var symbol = ConvertProductId(message.ProductId);
+
+                OrderBook orderBook;
+                if (!_orderBooks.TryGetValue(symbol, out orderBook))
+                {
+                    orderBook = new OrderBook(symbol);
+                    _orderBooks[symbol] = orderBook;
+                }
+                else
+                {
+                    orderBook.BestBidAskUpdated -= OnBestBidAskUpdated;
+                    orderBook.Clear();
+                }
+
+                foreach (var row in message.Bids)
+                {
+                    var price = decimal.Parse(row[0], NumberStyles.Float, CultureInfo.InvariantCulture);
+                    var size = decimal.Parse(row[1], NumberStyles.Float, CultureInfo.InvariantCulture);
+                    orderBook.UpdateBidRow(price, size);
+                }
+                foreach (var row in message.Asks)
+                {
+                    var price = decimal.Parse(row[0], NumberStyles.Float, CultureInfo.InvariantCulture);
+                    var size = decimal.Parse(row[1], NumberStyles.Float, CultureInfo.InvariantCulture);
+                    orderBook.UpdateAskRow(price, size);
+                }
+
+                orderBook.BestBidAskUpdated += OnBestBidAskUpdated;
+
+                EmitQuoteTick(symbol, orderBook.BestBidPrice, orderBook.BestBidSize, orderBook.BestAskPrice, orderBook.BestAskSize);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private void OnBestBidAskUpdated(object sender, BestBidAskUpdatedEventArgs e)
+        {
+            EmitQuoteTick(e.Symbol, e.BestBidPrice, e.BestBidSize, e.BestAskPrice, e.BestAskSize);
+        }
+
+        private void OnL2Update(string data)
+        {
+            try
+            {
+                var message = JsonConvert.DeserializeObject<Messages.L2Update>(data);
+
+                var symbol = ConvertProductId(message.ProductId);
+
+                var orderBook = _orderBooks[symbol];
```

Comment:
Is it possible for the order book to not be created yet? Does GDAX documentation define a certain order of messages, regarding snapshot vs L2 update?

Reply:
According to [GDAX docs for level2](https://docs.gdax.com/#the-code-classprettyprintlevel2code-channel), a _snapshot_ message will be sent after subscription, followed by _l2update_ messages.